File: ims-web/src/settings/printSettingsService.ts
export type PrintPaperFormat = 'A4' | 'A5' | 'A3' | 'Custom';

export interface SalesOrderPrintSettings {
  paperFormat: PrintPaperFormat;
  customWidthMm: number;
  customHeightMm: number;
  marginMm: number;
}

const STORAGE_KEY = 'ims.salesOrderPrint';

export const DEFAULT_PRINT_SETTINGS: SalesOrderPrintSettings = {
  paperFormat: 'A4',
  customWidthMm: 210,
  customHeightMm: 148,
  marginMm: 10,
};

function clampMm(value: number, min: number, max: number): number {
  if (!Number.isFinite(value)) return min;
  return Math.min(max, Math.max(min, value));
}

function normalize(raw: Partial<SalesOrderPrintSettings> | null | undefined): SalesOrderPrintSettings {
  const base = DEFAULT_PRINT_SETTINGS;
  const format = raw?.paperFormat;
  const paperFormat: PrintPaperFormat =
    format === 'A5' || format === 'A3' || format === 'Custom' ? format : 'A4';

  return {
    paperFormat,
    customWidthMm: clampMm(Number(raw?.customWidthMm ?? base.customWidthMm), 50, 1200),
    customHeightMm: clampMm(Number(raw?.customHeightMm ?? base.customHeightMm), 50, 1200),
    marginMm: clampMm(Number(raw?.marginMm ?? base.marginMm), 0, 50),
  };
}

export function getPageSizeMm(settings: SalesOrderPrintSettings): { widthMm: number; heightMm: number } {
  switch (settings.paperFormat) {
    case 'A5':
      return { widthMm: 148, heightMm: 210 };
    case 'A3':
      return { widthMm: 297, heightMm: 420 };
    case 'Custom':
      return { widthMm: settings.customWidthMm, heightMm: settings.customHeightMm };
    case 'A4':
    default:
      return { widthMm: 210, heightMm: 297 };
  }
}

export function formatPrintSummary(settings: SalesOrderPrintSettings): string {
  const { widthMm, heightMm } = getPageSizeMm(settings);
  const name = settings.paperFormat === 'Custom' ? 'Custom' : settings.paperFormat;
  return `${name} — ${widthMm.toFixed(1)} × ${heightMm.toFixed(1)} mm (portrait)`;
}

export function loadPrintSettings(): SalesOrderPrintSettings {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return { ...DEFAULT_PRINT_SETTINGS };
    return normalize(JSON.parse(raw) as Partial<SalesOrderPrintSettings>);
  } catch {
    return { ...DEFAULT_PRINT_SETTINGS };
  }
}

export function savePrintSettings(settings: SalesOrderPrintSettings): SalesOrderPrintSettings {
  const normalized = normalize(settings);
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(normalized));
  } catch {
    // ignore
  }
  return normalized;
}

File: IMS/Services/PrintSettingsService.cs
using System.Windows;
using IMS.Models;

namespace IMS.Services;

public sealed class PrintSettingsService
{
    private static readonly Lazy<PrintSettingsService> Lazy = new(() => new PrintSettingsService());
    public static PrintSettingsService Instance => Lazy.Value;

    private PrintSettingsService()
    {
        Current = new SalesOrderPrintSettings();
    }

    public SalesOrderPrintSettings Current { get; private set; }

    public event EventHandler? PrintSettingsChanged;

    public static void Initialize()
    {
        var settings = SettingsStore.Load();
        Instance.Apply(settings.SalesOrderPrint, persist: false);
    }

    public void Apply(SalesOrderPrintSettings settings, bool persist = true)
    {
        Current = new SalesOrderPrintSettings
        {
            PaperFormat = settings.PaperFormat,
            CustomWidthMm = ClampMm(settings.CustomWidthMm, 50, 1200),
            CustomHeightMm = ClampMm(settings.CustomHeightMm, 50, 1200),
            MarginMm = ClampMm(settings.MarginMm, 0, 50)
        };

        if (persist)
            SettingsStore.Update(s => s.SalesOrderPrint = Clone(Current));

        PrintSettingsChanged?.Invoke(this, EventArgs.Empty);
    }

    public (double WidthMm, double HeightMm) GetPageSizeMm() =>
        Current.PaperFormat switch
        {
            PrintPaperFormat.A4 => (210, 297),
            PrintPaperFormat.A5 => (148, 210),
            PrintPaperFormat.A3 => (297, 420),
            PrintPaperFormat.Custom => (Current.CustomWidthMm, Current.CustomHeightMm),
            _ => (210, 297)
        };

    public Size GetPageSizeDips()
    {
        var (w, h) = GetPageSizeMm();
        return new Size(MmToDip(w), MmToDip(h));
    }

    public Thickness GetPagePaddingDips() =>
        new(MmToDip(Current.MarginMm), MmToDip(Current.MarginMm), MmToDip(Current.MarginMm), MmToDip(Current.MarginMm));

    public PrintPageLayout GetPageLayout()
    {
        var pageSize = GetPageSizeDips();
        var padding = GetPagePaddingDips();
        var contentWidth = Math.Max(120, pageSize.Width - padding.Left - padding.Right);
        var (widthMm, heightMm) = GetPageSizeMm();
        return new PrintPageLayout(
            pageSize,
            padding,
            contentWidth,
            FormatDisplayName,
            $"{widthMm:0.#} × {heightMm:0.#} mm");
    }

    public string FormatDisplayName => Current.PaperFormat switch
    {
        PrintPaperFormat.A4 => "A4",
        PrintPaperFormat.A5 => "A5",
        PrintPaperFormat.A3 => "A3",
        PrintPaperFormat.Custom => "Custom",
        _ => "A4"
    };

    public string PageSizeSummary
    {
        get
        {
            var (w, h) = GetPageSizeMm();
            return $"{FormatDisplayName} — {w:0.#} × {h:0.#} mm (portrait)";
        }
    }

    public static double MmToDip(double mm) => mm * 96.0 / 25.4;

    public static double DipToMm(double dip) => dip * 25.4 / 96.0;

    private static double ClampMm(double value, double min, double max) =>
        Math.Clamp(value, min, max);

    private static SalesOrderPrintSettings Clone(SalesOrderPrintSettings s) => new()
    {
        PaperFormat = s.PaperFormat,
        CustomWidthMm = s.CustomWidthMm,
        CustomHeightMm = s.CustomHeightMm,
        MarginMm = s.MarginMm
    };
}

public readonly record struct PrintPageLayout(
    Size PageSizeDips,
    Thickness PagePaddingDips,
    double ContentWidthDips,
    string FormatName,
    string SizeLabel);
